fireapp/lib/domain/repository/notification_fcm_tokens_repository.dart

```diff
@@ -0,0 +1,67 @@
+import 'package:fireapp/data/client/fcm_tokens_client.dart';
+import 'package:fireapp/domain/models/fcm_tokens.dart';
+import 'package:injectable/injectable.dart';
+import 'package:firebase_messaging/firebase_messaging.dart';
+import 'dart:io';
+
+
+@injectable
+class NotificationFCMTokensRepository {
+  final FCMTokensClient tokensClient;
+  NotificationFCMTokensRepository(this.tokensClient);
+
+  Future<void> registerFCMTokens(int userId) async {
+    String? fcmToken = await FirebaseMessaging.instance.getToken();
+    if (fcmToken == null || fcmToken.isEmpty) {
+      throw Exception("FCM Token is null or empty.");
+    }
+
+    String deviceType;
+    if (Platform.isAndroid) {
+      deviceType = 'android';
+    } else if (Platform.isIOS) {
+      deviceType = 'ios';
+    } else {
+      deviceType = 'other';
+    }
+
+    DateTime creationTime = DateTime.now();
+    DateTime updateTime = DateTime.now();
+    int isActive = 1;
+
+    await tokensClient.registerFCMTokens(
+      FCMTokens(
+        userId: userId,
+        fcmToken: fcmToken,
+        deviceType: deviceType,
+        creationTime: creationTime,
+        updateTime: updateTime,
+        isActive: isActive,
+      ),
+    );
+    FirebaseMessaging.instance.onTokenRefresh.listen((newToken) async {
```

Comment:
Hi @BenMMcLean, thank you so much for the detailed feedback! I understand the bug you mentioned, but I'm not entirely sure about the use case, as MyApp doesn't interact with the repository. It makes sense to me to move the on-refresh token listener to the initState in MyApp, and let the LoginViewModel be responsible for posting the new FCM token to the database. Could you explain further how I should implement the use case here?

Reply:
The code in MyApp _should be_ interacting with the repository. As a rule, code above the data layer (Domain classes like ViewModels and UseCases) should _never_ interact with client and persistence classes. It may not be obvious from our fairly simple usage of the pattern in the app, but the purpose of having a repository layer is to abstract away data fetching and/or storage. An example of this you can view would be the `AuthenticationRepository`. In it, it uses the client to login, and then transparently caches the details in the persistence class. This leaves the domain layer to be more purely focused on business logic, rather than the peculiarities of data retrieval and storage.

The reason I suggest a use case is because there are two different ways you are interacting with the repository - one where you already have a token (refresh), and one where you don't (login). The repository should expect a token to be provided to it, and then you pull the logic to retrieve a token out of the repository and put it into a use case. To be clear, **you should update the repository function to accept an FCM token**. You can think of a use case as a discrete piece of business logic. Then the LoginViewModel just has to call an injected instance of the use case, rather than managing it itself.

In terms of implementing it, just fill in the code I've written above (read the link to call operator if you don't understand invoking an object instance). It should just be a matter of getting the token, then calling the repository function with that token. Then you just need to inject it into `LoginViewModel` and invoke it where you were previously calling the repository function.

Reply:
Hi @BenMMcLean,

Thank you for your detailed explanation, and I fully understand it. However, I am currently facing a problem where I tried to listen to token refresh from MyApp and update the database by calling a backend endpoint, but it's requiring both UserId and FCM token. The token is always available, but the UserId is not available at startup—it's only available after sign-in.

I can think of a solution to implement a simple fix in the backend endpoint or perform a null pointer check inside the lambda. Overall, I think the token refresh callback is not that important, as it is rarely triggered—only in extreme cases according to the documentation I found.

Reply:
It's true it is only triggered in extreme cases, but we still need to implement it. Just check if `AuthenticationRepository.getCurrentSession()` is null or not in the lambda. Make sure the lambda is in a ViewModel though, more broadly it would be useful to have an application scoped viewmodel

Reply:
The `AuthenticationRepository.getCurrentSession()` will only return a non-null value if the user is already logged in. This is because `getCurrentSession()` caches the session after the` logIn` function is triggered. Therefore, there are still cases where the user might not be logged in when the token refreshes.

The best approach I can think of is to use `getCurrentSession()` and only update the database through the backend if it returns a non-null value—meaning, only if the user is logged in. This logic makes sense to me.

What I plan to do next is create a new ViewModel class for main.dart to handle interactions with the FCM repository. I believe I'll need to delete the old token and add the new one during the token refresh process.

Please let me know if I'm on the right track! Thank you so much!

Reply:
Yes that is broadly what I was suggesting. You do not need to delete the old token, the backend will know if it is invalid when it attempts to send a notification.